File: EcSignatureP256Sha256DerString/EcSignatureDerStringFull.cs
using System;
using Org.BouncyCastle.Asn1.X9;
using Org.BouncyCastle.Security;

public class EcSignatureString {
	public static void Main() {
		// running with .NET 4.7.2 and BouncyCastle 1.89

		Console.WriteLine("EC signature string (ECDSA with SHA256) DER-encoding");
		var dataToSignString = "The quick brown fox jumps over the lazy dog";
		var dataToSign = System.Text.Encoding.UTF8.GetBytes(dataToSignString);
		
		Console.WriteLine("dataToSign: " + dataToSignString);
		try {
			Console.WriteLine("\n* * * sign the plaintext with the EC private key * * *");
			string ecPrivateKeyPem = loadEcPrivateKeyPem();
			string signatureBase64 = ecSignDerToBase64(ecPrivateKeyPem, dataToSign);
			Console.WriteLine("used private key:");
			Console.WriteLine(ecPrivateKeyPem);
			Console.WriteLine("\nsignature (Base64): " + signatureBase64);	
			
			// verify
			Console.WriteLine("\n* * * verify the signature against hash of plaintext with the EC public key * * *");
			string signatureReceivedBase64 = signatureBase64;
			string ecPublicKeyPem = loadEcPublicKeyPem();
			Console.WriteLine("used public key:");
			Console.WriteLine(ecPublicKeyPem);
			//bool signatureVerified = false;
			bool signatureVerified = ecVerifySignatureDerFromBase64(ecPublicKeyPem, dataToSign, signatureReceivedBase64);
			Console.WriteLine("\nsignature verified: " + signatureVerified);
		}
		catch(ArgumentNullException) {
			Console.WriteLine("The data was not signed or verified");
		}
	}
	
	private static string ecSignDerToBase64(string ecPrivateKeyPem, byte[] messageByte) {
		var privateKey = PrivateKeyFactory.CreateKey(Convert.FromBase64String(getEcPrivateKeyFromPemStripped(ecPrivateKeyPem)));
		var signer = SignerUtilities.GetSigner(X9ObjectIdentifiers.ECDsaWithSha256.Id);
		signer.Init(true, privateKey);
		signer.BlockUpdate(messageByte, 0, messageByte.Length);
		var signature = signer.GenerateSignature();
		return Base64Encoding(signature);
	}	
	
	private static bool ecVerifySignatureDerFromBase64(string ecPublicKeyPem, byte[] messageByte, String signatureBase64) {
		var publicKey = PublicKeyFactory.CreateKey(Convert.FromBase64String(getEcPublicKeyFromPemStripped(ecPublicKeyPem)));
		var verifier = SignerUtilities.GetSigner(X9ObjectIdentifiers.ECDsaWithSha256.Id);
		verifier.Init(false, publicKey);
		verifier.BlockUpdate(messageByte, 0, messageByte.Length);
		return verifier.VerifySignature(Convert.FromBase64String(signatureBase64));
	}
		
	private static string getEcPrivateKeyFromPemStripped(string ecPrivateKeyPem) {
		string ecPrivateKeyHeaderPem = "-----BEGIN PRIVATE KEY-----\n";
		string ecPrivateKeyFooterPem = "-----END PRIVATE KEY-----";
		string ecPrivateKeyDataPem = ecPrivateKeyPem.Replace(ecPrivateKeyHeaderPem, "").Replace(ecPrivateKeyFooterPem, "").Replace("\n", "");
		return ecPrivateKeyDataPem;
	}
	
	private static string getEcPublicKeyFromPemStripped(string ecPublicKeyPem) {
		string ecPublicKeyHeaderPem = "-----BEGIN PUBLIC KEY-----\n";
		string ecPublicKeyFooterPem = "-----END PUBLIC KEY-----";
		string ecPublicKeyDataPem = ecPublicKeyPem.Replace(ecPublicKeyHeaderPem, "").Replace(ecPublicKeyFooterPem, "").Replace("\n", "");
		return ecPublicKeyDataPem;
	}
	
	private static byte[] getEcPrivateKeyEncodedFromPem(string ecPrivateKeyPem) {
		string ecPrivateKeyHeaderPem = "-----BEGIN PRIVATE KEY-----\n";
		string ecPrivateKeyFooterPem = "-----END PRIVATE KEY-----";
		string ecPrivateKeyDataPem = ecPrivateKeyPem.Replace(ecPrivateKeyHeaderPem, "").Replace(ecPrivateKeyFooterPem, "").Replace("\n", "");
		return Base64Decoding(ecPrivateKeyDataPem);
	}

	private static byte[] getEcPublicKeyFromPem(string ecPublicKeyPem) {
		string ecPublicKeyHeaderPem = "-----BEGIN PUBLIC KEY-----\n";
		string ecPublicKeyFooterPem = "-----END PUBLIC KEY-----";
		string ecPublicKeyDataPem = ecPublicKeyPem.Replace(ecPublicKeyHeaderPem, "").Replace(ecPublicKeyFooterPem, "").Replace("\n", "");
		return Base64Decoding(ecPublicKeyDataPem);
	}

	private static string loadEcPrivateKeyPem() {
		// this is a sample key - don't worry !
		return "-----BEGIN PRIVATE KEY-----\n"
			+ "MIGHAgEAMBMGByqGSM49AgEGCCqGSM49AwEHBG0wawIBAQQgciPt/gulzw7/Xe12\n"  
			+ "YOu/vLUgIUZ+7gGo5VkmU0B+gUWhRANCAAQxkee3UPW110s0aUQdcS0TDkr8blAe\n"
			+ "SBouL4hXziiJX5Me/8OobFgNfYXkk6R/K/fqJhJ/mV8gLur16XhgueXA\n"
			+ "-----END PRIVATE KEY-----";
	}

	private static string loadEcPublicKeyPem() {
		// this is a sample key - don't worry !
		return "-----BEGIN PUBLIC KEY-----\n"
			+ "MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAEMZHnt1D1tddLNGlEHXEtEw5K/G5Q\n"
			+ "HkgaLi+IV84oiV+THv/DqGxYDX2F5JOkfyv36iYSf5lfIC7q9el4YLnlwA==\n"
			+ "-----END PUBLIC KEY-----";
	}

	static string Base64Encoding(byte[] input) {
		return Convert.ToBase64String(input);
	}

	static byte[] Base64Decoding(String input) {
		return Convert.FromBase64String(input);
	}
}

File: EcSignatureP256Sha256DerString/EcSignatureDerStringFull.py
from Crypto.Hash import SHA256
from Crypto.PublicKey import ECC
from Crypto.Signature import DSS
import base64

def base64Encoding(input):
  dataBase64 = base64.b64encode(input)
  dataBase64P = dataBase64.decode("UTF-8")
  return dataBase64P

def base64Decoding(input):
    return base64.decodebytes(input.encode("ascii"))

def ecSignToBase64(privateKeyPem, message):
  ecPrivateKey = ECC.import_key(privateKeyPem)
  hashMessage = SHA256.new(message.encode("ascii"))
  signer = DSS.new(ecPrivateKey, 'fips-186-3', 'der')
  signature = signer.sign(hashMessage)
  return base64Encoding(signature)

def ecVerifySignatureFromBase64(publicKeyPem, message, signatureBase64):
  ecPublicKey = ECC.import_key(publicKeyPem)
  signatureBytes = base64Decoding(signatureBase64)
  hashMessage = SHA256.new(message.encode("ascii"))
  verifier = DSS.new(ecPublicKey, 'fips-186-3', 'der')
  try:
    verifier.verify(hashMessage, signatureBytes)
    return True
  except ValueError:
    return False

def loadEcPrivateKeyPem():
  return """-----BEGIN PRIVATE KEY-----
MIGHAgEAMBMGByqGSM49AgEGCCqGSM49AwEHBG0wawIBAQQgciPt/gulzw7/Xe12
YOu/vLUgIUZ+7gGo5VkmU0B+gUWhRANCAAQxkee3UPW110s0aUQdcS0TDkr8blAe
SBouL4hXziiJX5Me/8OobFgNfYXkk6R/K/fqJhJ/mV8gLur16XhgueXA
-----END PRIVATE KEY-----
"""

def loadEcPublicKeyPem():
  return """-----BEGIN PUBLIC KEY-----
MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAEMZHnt1D1tddLNGlEHXEtEw5K/G5Q
HkgaLi+IV84oiV+THv/DqGxYDX2F5JOkfyv36iYSf5lfIC7q9el4YLnlwA==
-----END PUBLIC KEY-----"""

print("EC signature string (ECDSA with SHA256) DER-encoding")

dataToSign = "The quick brown fox jumps over the lazy dog"
print("dataToSign: " + dataToSign)

print("\n* * * sign the plaintext with the EC private key * * *")
privateEcKeyPem = loadEcPrivateKeyPem()
print("used private key:\n" + privateEcKeyPem)
signatureBase64 = ecSignToBase64(privateEcKeyPem, dataToSign)
print("signature (Base64): " + signatureBase64)

print("\n* * * verify the signature against the plaintext with the EC public key * * *")
publicEcKeyPem = loadEcPublicKeyPem()
print("used public key:\n" + publicEcKeyPem)
signatureVerified = ecVerifySignatureFromBase64(publicEcKeyPem, dataToSign, signatureBase64)
print("\nsignature (Base64) verified: ")
print(signatureVerified)
